posthog/async_migrations/migrations/0001_events_sample_by.py

```diff
@@ -0,0 +1,163 @@
+from ee.clickhouse.sql.events import EVENTS_TABLE
+from posthog.async_migrations.definition import AsyncMigrationDefinition, AsyncMigrationOperation
+from posthog.constants import AnalyticsDBMS
+from posthog.settings import CLICKHOUSE_CLUSTER, CLICKHOUSE_DATABASE
+from posthog.version_requirement import ServiceVersionRequirement
+
+TEMPORARY_TABLE_NAME = f"{CLICKHOUSE_DATABASE}.temp_events_0001_events_sample_by"
+EVENTS_TABLE_NAME = f"{CLICKHOUSE_DATABASE}.{EVENTS_TABLE}"
+BACKUP_TABLE_NAME = f"{EVENTS_TABLE_NAME}_backup_0001_events_sample_by"
+
+"""
+Migration Summary
+- Context: https://github.com/PostHog/posthog/issues/5684
+- Operations:
+    0. Create a new table with the updated schema: `SAMPLE BY cityHash64(distinct_id)` + `ORDER BY (team_id, toDate(timestamp), cityHash64(distinct_id), cityHash64(uuid))`
+    1. Start backfilling the new table (online) with data from partitions that are unlikely to be getting inserts (previous month and under)
+    2. Detach the events_mv materialized view so we stop ingestion
+    3. Insert the remaining events into the new table
+    4. Rename the current table to `events_backup_0001_events_sample_by` and rename the new table to `events` (the table we use for querying)
+    5. Attach the materialized view so we start ingestion again
+    6. Optimize the table to remove duplicates
+- Checks:
+    0. is_required: only run this on instances with the old schema (new deploys get the new schema by default)
+    1. precheck: make sure there's enough free disk space in CH to run the migration
+    2. healthcheck: prevent CH from blowing up for lack of disk space
+"""
+
+
+class Migration(AsyncMigrationDefinition):
+
+    description = (
+        "Schema change to the events table ensuring our SAMPLE BY clause is compatible with ClickHouse >=21.7.0."
+    )
+
+    posthog_min_version = "1.30.0"
+    posthog_max_version = "1.31.0"
+
+    service_version_requirements = [
+        ServiceVersionRequirement(service="clickhouse", supported_version=">=21.6.0,<21.7.0"),
+    ]
+
+    operations = [
+        AsyncMigrationOperation(
+            database=AnalyticsDBMS.CLICKHOUSE,
+            sql=f"""
+            CREATE TABLE IF NOT EXISTS {TEMPORARY_TABLE_NAME} ON CLUSTER {CLICKHOUSE_CLUSTER} AS {EVENTS_TABLE_NAME}
+            ENGINE = ReplacingMergeTree(_timestamp)
+            PARTITION BY toYYYYMM(timestamp)
+            ORDER BY (team_id, toDate(timestamp), cityHash64(distinct_id), cityHash64(uuid))
+            SAMPLE BY cityHash64(distinct_id) 
+            """,
+            rollback=f"DROP TABLE IF EXISTS {TEMPORARY_TABLE_NAME} ON CLUSTER {CLICKHOUSE_CLUSTER}",
+            resumable=True,
+        ),
+        AsyncMigrationOperation(
+            database=AnalyticsDBMS.CLICKHOUSE,
+            sql=f"""
+            INSERT INTO {TEMPORARY_TABLE_NAME}
+            SELECT * 
+            FROM {EVENTS_TABLE}
+            WHERE timestamp < toDateTime64(toYYYYMM(now()) - 1, 6)
```

Comment:
@fuziontech without the casting this would have failed. good thing I decided to add more tests before shipping :D